File: src/backend/polaris/cells/runtime/projection/public/role_contracts.py
"""Shared public role contracts for PM, Chief Engineer, and Director views."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field

ROLE_TASK_STATUS_VALUES: tuple[str, ...] = (
    "PENDING",
    "CLAIMED",
    "RUNNING",
    "BLOCKED",
    "FAILED",
    "COMPLETED",
    "CANCELLED",
)

ROLE_TASK_PRIORITY_VALUES: tuple[str, ...] = (
    "LOW",
    "MEDIUM",
    "HIGH",
    "CRITICAL",
)


class RoleTaskContractV1(BaseModel):
    """Canonical task row consumed by PM/Director desktop workspaces."""

    model_config = ConfigDict(extra="forbid")

    id: str
    subject: str
    description: str = ""
    status: str
    priority: str
    claimed_by: str | None = None
    result: dict[str, Any] | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    goal: str = ""
    acceptance: list[str] = Field(default_factory=list)
    target_files: list[str] = Field(default_factory=list)
    dependencies: list[str] = Field(default_factory=list)
    current_file: str | None = None
    error: str | None = None
    worker: str | None = None
    pm_task_id: str | None = None
    blueprint_id: str | None = None
    blueprint_path: str | None = None
    runtime_blueprint_path: str | None = None


class ChiefEngineerBlueprintSummaryV1(BaseModel):
    """Canonical Chief Engineer blueprint list item."""

    model_config = ConfigDict(extra="forbid")

    blueprint_id: str
    title: str = ""
    summary: str = ""
    status: str | None = None
    source: str = "runtime/blueprints"
    target_files: list[str] = Field(default_factory=list)
    updated_at: str | None = None
    raw: dict[str, Any] = Field(default_factory=dict)


class ChiefEngineerBlueprintListV1(BaseModel):
    """Canonical Chief Engineer blueprint list response."""

    model_config = ConfigDict(extra="forbid")

    blueprints: list[ChiefEngineerBlueprintSummaryV1]
    total: int


class ChiefEngineerBlueprintDetailV1(BaseModel):
    """Canonical Chief Engineer blueprint detail response."""

    model_config = ConfigDict(extra="forbid")

    blueprint_id: str
    source: str = "runtime/blueprints"
    blueprint: dict[str, Any]


__all__ = [
    "ROLE_TASK_PRIORITY_VALUES",
    "ROLE_TASK_STATUS_VALUES",
    "ChiefEngineerBlueprintDetailV1",
    "ChiefEngineerBlueprintListV1",
    "ChiefEngineerBlueprintSummaryV1",
    "RoleTaskContractV1",
]

File: src/frontend/src/types/roleContracts.ts
export const ROLE_TASK_STATUS_VALUES = [
  'PENDING',
  'CLAIMED',
  'RUNNING',
  'BLOCKED',
  'FAILED',
  'COMPLETED',
  'CANCELLED',
] as const;

export const ROLE_TASK_PRIORITY_VALUES = ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL'] as const;

export type RoleTaskStatusV1 = (typeof ROLE_TASK_STATUS_VALUES)[number];
export type RoleTaskPriorityV1 = (typeof ROLE_TASK_PRIORITY_VALUES)[number];

export interface RoleTaskContractV1 {
  id: string;
  subject: string;
  description: string;
  status: string;
  priority: string;
  claimed_by: string | null;
  result: Record<string, unknown> | null;
  metadata: Record<string, unknown>;
  goal: string;
  acceptance: string[];
  target_files: string[];
  dependencies: string[];
  current_file: string | null;
  error: string | null;
  worker: string | null;
  pm_task_id: string | null;
  blueprint_id: string | null;
  blueprint_path: string | null;
  runtime_blueprint_path: string | null;
}

export interface ChiefEngineerBlueprintSummaryV1 {
  blueprint_id: string;
  title: string;
  summary: string;
  status: string | null;
  source: string;
  target_files: string[];
  updated_at: string | null;
  raw: Record<string, unknown>;
}

export interface ChiefEngineerBlueprintListV1 {
  blueprints: ChiefEngineerBlueprintSummaryV1[];
  total: number;
}

export interface ChiefEngineerBlueprintDetailV1 {
  blueprint_id: string;
  source: string;
  blueprint: Record<string, unknown>;
}
